### DISCUSSION TITLE
Character movement issue

### DISCUSSION THREAD

**Discussion by long1752062104** [notgiven688/jitterphysics2#176] [How do I ... ?]
When I modified the player's movement speed, I found that the player gradually increased their force while walking, rather than moving smoothly
`Keyboard kb = Keyboard.Instance;

if (kb.IsKeyDown(Keyboard.Key.Left)) player.SetAngularInput(-1.0f);
else if (kb.IsKeyDown(Keyboard.Key.Right)) player.SetAngularInput(1.0f);
else player.SetAngularInput(0.0f);

if (kb.IsKeyDown(Keyboard.Key.Up)) player.SetLinearInput(-JVector.UnitZ * 20f);
else if (kb.IsKeyDown(Keyboard.Key.Down)) player.SetLinearInput(JVector.UnitZ * 20f);
else player.SetLinearInput(JVector.Zero);

if (kb.IsKeyDown(Keyboard.Key.LeftControl)) player.Jump();`

**Comment by notgiven688** [comment]
There is no character controller implementation in Jitter, just a demo scene within JitterDemo which can serve as a starting point for your own implementation. Have a look into Player.cs how the force is added. You can also implement your player as a "kinetic body", directly modifying the velocity of the rigid body.

**Reply by Alan-FGR** [reply]
Thanks for working on this awesome library @notgiven688.

It would be nice to have (perhaps as an "extension" kinda thing, or even an example) a simulation-independent character controller that uses the famous "move and slide" operation, and perhaps also one that allows pushing bodies or maybe attaching a rigid body so it affects the actual simulation.

Having a character controller dependent on the simulation has a lot of issues and can't provide the same level of responsivity and quality.

I am wondering if it would be viable to have a separate "simulation" only for the character controller logic, so we can enable CCD for some cylinders and move them around at rendering rate, while using the same scene from a fixed-rate "physics simulation".

**Reply by notgiven688** [reply]
@Alan-FGR 

> Thanks for working on this awesome library @notgiven688.

Thank You!

> It would be nice to have (perhaps as an "extension" kinda thing, or even an example) a simulation-independent character controller that uses the famous "move and slide" operation, and perhaps also one that allows pushing bodies or maybe attaching a rigid body so it affects the actual simulation.

You're absolutely right — a good starting point for a character controller would be helpful.

The challenge with character controllers is that they often need to be tailored to the specific needs of each game or application. Because of that, any built-in implementation would likely need customization, and a solid understanding of the underlying principles is required anyway. At that point, it can make sense for developers to roll their own.

This is actually part of the engine’s philosophy: Jitter aims to provide a solid and minimal foundation for building your own systems - whether it’s character controllers, vehicles, or something else - rather than making assumptions about how higher-level behaviors should work.

That said, I agree that having an example implementation would be useful. Do you have a reference implementation of the “move and slide” approach (link, paper, blog-post, ..)?

> I am wondering if it would be viable to have a separate "simulation" only for the character controller logic, so we can enable CCD for some cylinders and move them around at rendering rate, while using the same scene from a fixed-rate "physics simulation".

Yes, that’s definitely a viable approach. The CcdSolver implementation in the demo could be adapted to track "characters".

I am not even sure that this is needed. In fact, a good character controller might not rely on the full physics simulation at all — just on parts of it, like raycasts or sweep tests. That’s similar to how the "Ray-cast Car" demo works compared to the fully simulated "Constraint car" demo.

**Reply by Alan-FGR** [reply]
Regarding the "move and slide" (aka "collide and slide"), I think everybody still implements the [paper from Fauerby](https://www.peroxide.dk/papers/collision/collision.pdf). There's an [update to it though](https://arxiv.org/pdf/1211.0059).
In terms of articles, I could find this one which seems somewhat complete: [https://handmade.network/p/75/monter/blog/p/2916-collision_system_part_1__collision_detection_and_response_against_static_mesh](https://handmade.network/p/75/monter/blog/p/2916-collision_system_part_1__collision_detection_and_response_against_static_mesh).

I think the "move and slide" is not very different from the CCD with its sweep test and reprojection. It's actually a very simple concept.

I totally agree that a good character controller should work with raycasts and/or sweep tests, and should be somewhat of a scene query thing, which is totally independent from the simulation tick rate (it's not uncommon that games render at much faster rates than physics). Additionally the user can always sync a body in the simulation to the character controller in order to get callbacks and make the character have a "presence" in the scene for scene queries and such. This approach is built into Jolt and Rapier for example, which also implement the "move and slide", along with PhysX - in this latter the character controller is actually an extension on top of the base library. All of these projects are opensource (or at least "source available"), so maybe they could also serve as a reference for implementation.

**Reply by notgiven688** [reply]
I had a quick glance at the notes. They definitely cover the basics, but the devil is in the details—handling moving platforms, climbing stairs, navigating inclined planes, ducking, and so on. Looking at implementations like Jolt is especially interesting.
